### DISCUSSION TITLE
element.remove isn't work well with lit?

### DISCUSSION THREAD

**Discussion by dbwodlf3** [lit/lit#3014] [Q&A]
I tried to make some toast component.
my first idea is just use ```setTimeout``` with render function of lit to remove toast text element.
but render didn't work well. because i use just "this.remove". maybe lit ```render``` function reuse element.

I think this problem is coming from breaking context between lit and document nodes.
I don't know actually what is correct...

[code](https://lit.dev/playground/#project=W3sibmFtZSI6InNpbXBsZS1ncmVldGluZy50cyIsImNvbnRlbnQiOiJpbXBvcnQgeyBMaXRFbGVtZW50LCBodG1sLCBjc3MsIHJlbmRlciB9IGZyb20gJ2xpdCc7XG5pbXBvcnQgeyBjdXN0b21FbGVtZW50LCBwcm9wZXJ0eSwgcXVlcnkgfSBmcm9tICdsaXQvZGVjb3JhdG9ycy5qcyc7XG5cbkBjdXN0b21FbGVtZW50KCd0b2FzdC1jb21wb25lbnQnKVxuZXhwb3J0IGNsYXNzIFRvYXN0Q29tcG9uZW50IGV4dGVuZHMgTGl0RWxlbWVudCB7XG4gIF9pbm5lcklkPTA7XG4gIHRvYXN0SXRlbXM6IGFueSA9IFtdO1xuXG4gIHJlbmRlcigpe1xuICAgIHJldHVybiBodG1sYGA7XG4gIH1cblxuICBjb25uZWN0ZWRDYWxsYmFjaygpe1xuICAgIHN1cGVyLmNvbm5lY3RlZENhbGxiYWNrKCk7XG4gICAgd2luZG93LmFkZEV2ZW50TGlzdGVuZXIoXCJ0b2FzdF9tZXNzYWdlXCIsKGU6IGFueSk9PntcbiAgICAgIGNvbnN0IHR5cGUgPSBlLmRldGFpbC50eXBlO1xuICAgICAgY29uc3QgdGV4dCA9IGUuZGV0YWlsLnRleHQ7XG4gICAgICByZW5kZXIoaHRtbGA8dG9hc3QtaXRlbSB0eXBlPSR7dHlwZX0-JHt0ZXh0fTwvdG9hc3QtaXRlbT5gLCBkb2N1bWVudC5ib2R5KTtcblxuICAgICAgY29uc29sZS5sb2coXCJDQUxMIFRPQVNUIElURU0hISEhXCIpO1xuICAgIH0pO1xuICB9XG4gIFxufVxuXG5AY3VzdG9tRWxlbWVudCgndG9hc3QtaXRlbScpXG5leHBvcnQgY2xhc3MgVG9hc3RJdGVtIGV4dGVuZHMgTGl0RWxlbWVudCB7XG5cbiAgc3RhdGljIHN0eWxlcyA9IGNzc2BcblxuICAgIC5kZWZhdWx0LWNvbnRhaW5lciB7XG4gICAgICBwb3NpdGlvbjphYnNvbHVpdGU7XG4gICAgICB0b3A6IDUwcHg7XG4gICAgICBsZWZ0OiAyMDBweDtcbiAgICAgIHotaW5kZXg6IDk5OTk5O1xuICAgICAgYmFja2dyb3VuZDogIzAwMDtcbiAgICAgIGNvbG9yOiAjZmZmO1xuICAgICAgZGlzcGxheTogZmxleDsgcGFkZGluZzogNXB4IDEwcHggNXB4IDEwcHg7XG4gICAgfVxuXG4gICAgLmRlZmF1bHQtY29udGFpbmVyLnN1Y2Nlc3Mge1xuICAgICAgYm9yZGVyOiAxcHggc29saWQgIzVkYmJmYztcbiAgICB9XG4gIGA7XG5cbiAgQHByb3BlcnR5KClcbiAgdHlwZSA9IFwiaW5mb1wiO1xuXG4gIHJlbmRlcigpe1xuICAgIHJldHVybiBodG1sYFxuICAgICAgPGRpdiBjbGFzcz1cImRlZmF1bHQtY29udGFpbmVyICR7dGhpcy50eXBlfVwiPlxuICAgICAgICA8ZGl2IGNsYXNzPVwiJHt0aGlzLnR5cGV9LXRleHRcIj48c2xvdD48L3Nsb3Q-PC9kaXY-XG4gICAgICA8L2Rpdj5cbiAgICBgO1xuICB9XG5cbiAgZmlyc3RVcGRhdGVkKCl7XG4gICAgc2V0VGltZW91dCgoKT0-e1xuICAgICAgdGhpcy5yZW1vdmUoKTtcbiAgICB9LCAzMDAwKVxuXG4gICAgdGhpcy5zdHlsZS5wb3NpdGlvbiA9IFwiYWJzb2x1dGVcIjtcbiAgICB0aGlzLnN0eWxlLnRvcCA9IFwiMjBweFwiO1xuICAgIHRoaXMuc3R5bGUubGVmdCA9IFwiMjAwcHhcIjtcbiAgICB0aGlzLnN0eWxlLnpJbmRleCA9IFwiOTk5OVwiO1xuXG4gICAgY29uc29sZS5sb2coXCJSRU5ERVJFRCBUT0FTVCBJVEVNIVwiKTtcbiAgfVxufVxuXG5zZXRUaW1lb3V0KCgpPT57XG53aW5kb3cuZGlzcGF0Y2hFdmVudChuZXcgQ3VzdG9tRXZlbnQoXCJ0b2FzdF9tZXNzYWdlXCIsIHtkZXRhaWw6e3R5cGU6XCJpbmZvXCIsIHRleHQ6XCJIZWxsbyBNci4gVG9hc3RcIn19KSk7XG59LCAxMDAwKVxuc2V0VGltZW91dCgoKT0-e1xuICB3aW5kb3cuZGlzcGF0Y2hFdmVudChuZXcgQ3VzdG9tRXZlbnQoXCJ0b2FzdF9tZXNzYWdlXCIsIHtkZXRhaWw6e3R5cGU6XCJpbmZvXCIsIHRleHQ6XCJTZWNvbmQgSGVsbG8gTXIuIFRvYXN0XCJ9fSkpO1xufSwgNTAwMCkifSx7Im5hbWUiOiJpbmRleC5odG1sIiwiY29udGVudCI6IjwhRE9DVFlQRSBodG1sPlxuPGhlYWQ-XG4gIDxzY3JpcHQgdHlwZT1cIm1vZHVsZVwiIHNyYz1cIi4vc2ltcGxlLWdyZWV0aW5nLmpzXCI-PC9zY3JpcHQ-XG48L2hlYWQ-XG48Ym9keT5cbiAgPHRvYXN0LWNvbXBvbmVudCBuYW1lPVwiV29ybGRcIj48L3RvYXN0LWNvbXBvbmVudD5cbjwvYm9keT5cblxuPHNjcmlwdD4gPC9zY3JpcHQ-In0seyJuYW1lIjoicGFja2FnZS5qc29uIiwiY29udGVudCI6IntcbiAgXCJkZXBlbmRlbmNpZXNcIjoge1xuICAgIFwibGl0XCI6IFwiXjIuMC4wXCIsXG4gICAgXCJAbGl0L3JlYWN0aXZlLWVsZW1lbnRcIjogXCJeMS4wLjBcIixcbiAgICBcImxpdC1lbGVtZW50XCI6IFwiXjMuMC4wXCIsXG4gICAgXCJsaXQtaHRtbFwiOiBcIl4yLjAuMFwiXG4gIH1cbn0iLCJoaWRkZW4iOnRydWV9XQ)

**Comment by ynxtynxt** [comment]
If you don't want to reuse the template then maybe you can associate it with keyed() directive?
```ts
      render(keyed(this._itemKey++,html`<toast-item type=${type}>${text}</toast-item>`), document.body);
```

[🛝🛹🥏](https://lit.dev/playground/#project=W3sibmFtZSI6InNpbXBsZS1ncmVldGluZy50cyIsImNvbnRlbnQiOiJpbXBvcnQgeyBMaXRFbGVtZW50LCBodG1sLCBjc3MsIHJlbmRlciB9IGZyb20gJ2xpdCc7XG5pbXBvcnQgeyBjdXN0b21FbGVtZW50LCBwcm9wZXJ0eSwgcXVlcnkgfSBmcm9tICdsaXQvZGVjb3JhdG9ycy5qcyc7XG5pbXBvcnQge2tleWVkfSBmcm9tICdsaXQvZGlyZWN0aXZlcy9rZXllZC5qcyc7XG5cbkBjdXN0b21FbGVtZW50KCd0b2FzdC1jb21wb25lbnQnKVxuZXhwb3J0IGNsYXNzIFRvYXN0Q29tcG9uZW50IGV4dGVuZHMgTGl0RWxlbWVudCB7XG4gIF9pbm5lcklkPTA7XG4gIHRvYXN0SXRlbXM6IGFueSA9IFtdO1xuXG4gIHJlbmRlcigpe1xuICAgIHJldHVybiBodG1sYGA7XG4gIH1cblxuICBfaXRlbUtleSA9IDA7XG4gIGNvbm5lY3RlZENhbGxiYWNrKCl7XG4gICAgc3VwZXIuY29ubmVjdGVkQ2FsbGJhY2soKTtcbiAgICB3aW5kb3cuYWRkRXZlbnRMaXN0ZW5lcihcInRvYXN0X21lc3NhZ2VcIiwoZTogYW55KT0-e1xuICAgICAgY29uc3QgdHlwZSA9IGUuZGV0YWlsLnR5cGU7XG4gICAgICBjb25zdCB0ZXh0ID0gZS5kZXRhaWwudGV4dDtcbiAgICAgIHJlbmRlcihrZXllZCh0aGlzLl9pdGVtS2V5KyssaHRtbGA8dG9hc3QtaXRlbSB0eXBlPSR7dHlwZX0-JHt0ZXh0fTwvdG9hc3QtaXRlbT5gKSwgZG9jdW1lbnQuYm9keSk7XG4gICAgIFxuXG4gICAgICBjb25zb2xlLmxvZyhcIkNBTEwgVE9BU1QgSVRFTSEhISFcIik7XG4gICAgfSk7XG4gIH1cbiAgXG59XG5cbkBjdXN0b21FbGVtZW50KCd0b2FzdC1pdGVtJylcbmV4cG9ydCBjbGFzcyBUb2FzdEl0ZW0gZXh0ZW5kcyBMaXRFbGVtZW50IHtcblxuICBzdGF0aWMgc3R5bGVzID0gY3NzYFxuXG4gICAgLmRlZmF1bHQtY29udGFpbmVyIHtcbiAgICAgIHBvc2l0aW9uOmFic29sdWl0ZTtcbiAgICAgIHRvcDogNTBweDtcbiAgICAgIGxlZnQ6IDIwMHB4O1xuICAgICAgei1pbmRleDogOTk5OTk7XG4gICAgICBiYWNrZ3JvdW5kOiAjMDAwO1xuICAgICAgY29sb3I6ICNmZmY7XG4gICAgICBkaXNwbGF5OiBmbGV4OyBwYWRkaW5nOiA1cHggMTBweCA1cHggMTBweDtcbiAgICB9XG5cbiAgICAuZGVmYXVsdC1jb250YWluZXIuc3VjY2VzcyB7XG4gICAgICBib3JkZXI6IDFweCBzb2xpZCAjNWRiYmZjO1xuICAgIH1cbiAgYDtcblxuICBAcHJvcGVydHkoKVxuICB0eXBlID0gXCJpbmZvXCI7XG5cbiAgcmVuZGVyKCl7XG4gICAgcmV0dXJuIGh0bWxgXG4gICAgICA8ZGl2IGNsYXNzPVwiZGVmYXVsdC1jb250YWluZXIgJHt0aGlzLnR5cGV9XCI-XG4gICAgICAgIDxkaXYgY2xhc3M9XCIke3RoaXMudHlwZX0tdGV4dFwiPjxzbG90Pjwvc2xvdD48L2Rpdj5cbiAgICAgIDwvZGl2PlxuICAgIGA7XG4gIH1cblxuICBmaXJzdFVwZGF0ZWQoKXtcbiAgICBzZXRUaW1lb3V0KCgpPT57XG4gICAgICB0aGlzLnJlbW92ZSgpO1xuICAgIH0sIDMwMDApXG5cbiAgICB0aGlzLnN0eWxlLnBvc2l0aW9uID0gXCJhYnNvbHV0ZVwiO1xuICAgIHRoaXMuc3R5bGUudG9wID0gXCIyMHB4XCI7XG4gICAgdGhpcy5zdHlsZS5sZWZ0ID0gXCIyMDBweFwiO1xuICAgIHRoaXMuc3R5bGUuekluZGV4ID0gXCI5OTk5XCI7XG5cbiAgICBjb25zb2xlLmxvZyhcIlJFTkRFUkVEIFRPQVNUIElURU0hXCIpO1xuICB9XG59XG5cbnNldFRpbWVvdXQoKCk9PntcbndpbmRvdy5kaXNwYXRjaEV2ZW50KG5ldyBDdXN0b21FdmVudChcInRvYXN0X21lc3NhZ2VcIiwge2RldGFpbDp7dHlwZTpcImluZm9cIiwgdGV4dDpcIldlXCJ9fSkpO1xufSwgMTAwMClcbnNldFRpbWVvdXQoKCk9PntcbiAgd2luZG93LmRpc3BhdGNoRXZlbnQobmV3IEN1c3RvbUV2ZW50KFwidG9hc3RfbWVzc2FnZVwiLCB7ZGV0YWlsOnt0eXBlOlwiaW5mb1wiLCB0ZXh0Olwi8J-SnlwifX0pKTtcbn0sIDUwMDApXG5zZXRUaW1lb3V0KCgpPT57XG4gIHdpbmRvdy5kaXNwYXRjaEV2ZW50KG5ldyBDdXN0b21FdmVudChcInRvYXN0X21lc3NhZ2VcIiwge2RldGFpbDp7dHlwZTpcImluZm9cIiwgdGV4dDpcIkxpdCFcIn19KSk7XG59LCAxMDAwMCkifSx7Im5hbWUiOiJpbmRleC5odG1sIiwiY29udGVudCI6IjwhRE9DVFlQRSBodG1sPlxuPGhlYWQ-XG4gIDxzY3JpcHQgdHlwZT1cIm1vZHVsZVwiIHNyYz1cIi4vc2ltcGxlLWdyZWV0aW5nLmpzXCI-PC9zY3JpcHQ-XG48L2hlYWQ-XG48Ym9keT5cbiAgPHRvYXN0LWNvbXBvbmVudCBuYW1lPVwiV29ybGRcIj48L3RvYXN0LWNvbXBvbmVudD5cbjwvYm9keT5cblxuPHNjcmlwdD4gPC9zY3JpcHQ-In0seyJuYW1lIjoicGFja2FnZS5qc29uIiwiY29udGVudCI6IntcbiAgXCJkZXBlbmRlbmNpZXNcIjoge1xuICAgIFwibGl0XCI6IFwiXjIuMC4wXCIsXG4gICAgXCJAbGl0L3JlYWN0aXZlLWVsZW1lbnRcIjogXCJeMS4wLjBcIixcbiAgICBcImxpdC1lbGVtZW50XCI6IFwiXjMuMC4wXCIsXG4gICAgXCJsaXQtaHRtbFwiOiBcIl4yLjAuMFwiXG4gIH1cbn0iLCJoaWRkZW4iOnRydWV9XQ)

**Reply by dbwodlf3** [reply]
Yep. it can solve this, but it seems tricky. I just solve this by the other ways. i am just curious there are the natural way.

**Reply by ynxtynxt** [reply]
A more natural way would be manipulating your own DOM? (At least, I did that all the time.)

**Reply by dbwodlf3** [reply]
My first idea is just use element.remove(this.remove, in lit element) without anything. then naturally it is also reflected to lit, but not.

So I just it is push to parents array (for rendering toast item) and delete array with requestUpdate.

**Reply by ynxtynxt** [reply]
Maybe the item shouldn't remove itself but maintaining their visibility state within the parent element?

Yes, I guess I'd do it that way.

BTW, I have a chance to make my toast recently.. and I pop all my toasts up with asyncAppend() and I don't see a problem yet.

[gist](https://gist.github.com/SaladHut/222fd9694601412736ff2829c64b98dc)
